### PR TITLE
Intent picker

### PR DESCRIPTION
**Pull request by kevc**
@pardom 

The back traversal is still a bit goofy. I have most of the work done for showing a self-rolled bottom sheet picker, but it got kind of weird and I wanted to see what you thought of this first.

### CODE REVIEW

File: android/src/main/kotlin/clean/news/ui/item/share/ItemShareIntent.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import dagger.Module
+import dagger.Provides
+import dagger.Subcomponent
+import flow.ClassKey
+import flow.TreeKey
+import nz.bradcampbell.paperparcel.PaperParcel
+import nz.bradcampbell.paperparcel.PaperParcelable
+
+@PaperParcel
+class ItemShareIntent(val item: Item) : ClassKey(),
```
Comment (pardom-zz) [inline android/src/main/kotlin/clean/news/ui/item/share/ItemShareIntent.kt]: `Intent` is not platform agnostic. I'd come up with some more generic name, e.g. `ItemShareScreen` or `ItemShareModal`.

File: core/src/main/kotlin/clean/news/core/entity/Item.kt [hunk 2]
```diff
@@ -30,4 +30,6 @@ data class Item(
 	companion object {
 		val EMPTY = Item(-1, null, Type.NONE, null, Date(0), null, null, null, null, null, null, null, null, null, -1)
 	}
+
+	fun threadUrl(): String = "https://news.ycombinator.com/item?id=${this.id}"
```
Comment (pardom-zz) [inline core/src/main/kotlin/clean/news/core/entity/Item.kt]: This doesn't belong in `core`. It _probably_ belongs in `data` in the `retrofit` directory as a util object or something.

File: android/src/main/kotlin/clean/news/ui/item/share/ItemShareIntent.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		TreeKey,
+		ItemShareKey,
+		WithIntent,
+		WithComponent<ItemDetailComponent>,
+		PaperParcelable {
+
+	override fun getParentKey() = ItemDetailScreen(item)
+
+	override fun createComponent(parent: ItemDetailComponent) = parent.plus(
```
Comment (pardom-zz) [inline android/src/main/kotlin/clean/news/ui/item/share/ItemShareIntent.kt]: This formatting is odd. Are you using 120 (i.e. the Notion config) for this project? I'd prefer to keep single expression functions to one line.

File: android/src/main/res/menu/item_detail_view.xml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<menu xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (pardom-zz) [inline android/src/main/res/menu/item_detail_view.xml:2]: We should also add this menu item to the url view.

File: android/src/main/kotlin/clean/news/ui/item/share/ItemShareIntent.kt (hunk 1 above)
Comment (kevc) [inline android/src/main/kotlin/clean/news/ui/item/share/ItemShareIntent.kt]: It resides in the `android` module, though. Same thing with the `WithIntent` interface. Does it need to be platform agnostic?

File: android/src/main/kotlin/clean/news/ui/item/share/ItemShareIntent.kt (hunk 3 above)
Comment (kevc) [inline android/src/main/kotlin/clean/news/ui/item/share/ItemShareIntent.kt]: Looks like our codestyle has it set to 100, and this line goes to 102. Will remove the wrapping.

File: android/src/main/kotlin/clean/news/ui/item/share/ItemShareIntent.kt (hunk 1 above)
Comment (pardom-zz) [inline android/src/main/kotlin/clean/news/ui/item/share/ItemShareIntent.kt]: Yeah, I guess you're right about the platform agnostic point. However, I think the `Intent` descriptor is more if an implementation detail and is addressed by the `WithIntent` implementation. I think these key implementations should be named by what they are (screens) not how they accomplish their presentation (view/intent/etc). Thoughts?

**Comment by kevc** [comment]
Added sharing from URL screen, changed nomenclature to "Screen" for sharing.

File: android/src/main/kotlin/clean/news/ui/item/detail/ItemDetailView.kt [hunk 5]
```diff
@@ -46,17 +47,19 @@ class ItemDetailView : RelativeLayout {
...
 	override fun onAttachedToWindow() {
 		super.onAttachedToWindow()
-
 		val sinks = model.setUp(Sources(
 				toolbar.navigationClicks(),
-				Observable.empty()
+				Observable.empty(),
+				toolbar.itemClicks()
+						.filter {it.itemId == R.id.item_share }
+						.map { Unit }
```
Comment (pardom-zz) [inline android/src/main/kotlin/clean/news/ui/item/detail/ItemDetailView.kt:62]: I think if we change the source observable to `Observable<Any>`, we can avoid the mapping. Sorry,  I didn't catch this last round of comments.

File: android/src/main/kotlin/clean/news/ui/item/url/ItemUrlView.kt [hunk 6]
```diff
@@ -59,6 +59,9 @@ class ItemUrlView : RelativeLayout {
 				toolbar.navigationClicks(),
 				toolbar.itemClicks()
 						.filter { it.itemId == R.id.item_details }
+						.map { Unit },
```
Comment (pardom-zz) [inline android/src/main/kotlin/clean/news/ui/item/url/ItemUrlView.kt:62]: Use `Observable<Any>` to avoid mapping.

**Comment by pardom-zz** [comment]
LGTM. Make that mapping change and merge, or merge and fix have a look later. 

😴

**Comment by kevc** [comment]
The tap sources have to be typed as wildcards (`Observable<*>`). Using `Any` doesn't work with the parameterized type. We can decide what we want to do later.

**Pull request merged**